crates/hir-def/src/path/lower.rs

```diff
@@ -17,23 +17,58 @@ use crate::{
     type_ref::{LifetimeRef, TypeBound, TypeRef},
 };
 
+#[cfg(test)]
+thread_local! {
+    /// This is used to test `hir_segment_to_ast_segment()`. It's a hack, but it makes testing much easier.
+    pub(super) static SEGMENT_LOWERING_MAP: std::cell::RefCell<rustc_hash::FxHashMap<ast::PathSegment, usize>> = std::cell::RefCell::default();
+}
+
 /// Converts an `ast::Path` to `Path`. Works with use trees.
 /// It correctly handles `$crate` based path from macro call.
+// If you modify the logic of the lowering, make sure to check if `hir_segment_to_ast_segment()`
+// also needs an update.
 pub(super) fn lower_path(ctx: &mut LowerCtx<'_>, mut path: ast::Path) -> Option<Path> {
     let mut kind = PathKind::Plain;
     let mut type_anchor = None;
     let mut segments = Vec::new();
     let mut generic_args = Vec::new();
+    #[cfg(test)]
+    let mut ast_segments = Vec::new();
+    #[cfg(test)]
+    let mut ast_segments_offset = 0;
+    #[allow(unused_mut)]
+    let mut push_segment = |_segment: &ast::PathSegment, segments: &mut Vec<Name>, name| {
+        #[cfg(test)]
+        ast_segments.push(_segment.clone());
+        segments.push(name);
+    };
     loop {
         let segment = path.segment()?;
 
-        if segment.coloncolon_token().is_some() {
-            kind = PathKind::Abs;
+        if let Some(coloncolon) = segment.coloncolon_token() {
+            if coloncolon
+                .prev_token()
+                .is_none_or(|it| !path.syntax().text_range().contains_range(it.text_range()))
```

Comment:
`coloncolon == segment.first_token()` feels like a better check to me than wrangling ranges.

I also wonder if we should adjust our parse tree for this construct, as `::Fn::()` actually has a segment with two colon2 tokens which doesn't seem right. But that shouldn't concern this PR